### ISSUE TITLE
Clash of the debouncers

### ISSUE THREAD

**Issue by amsross** [regular/pull-debounce#1]
Rather than pollute https://github.com/pull-stream/pull-stream-docs/pull/27 with this discussion, I'm opening this issue.

I have a feeling that our two implementations of debounce may serve two different purposes.

Your implementation using `pull-window` will return the last data event in one window regardless of silence on the wire within or after one window. [My implementation](https://github.com/amsross/pull-debounce) will wait until there has been silence on the wire for the length of one window, at which time it will return the last data event. As a result, my implementation will not pass your tests, and your implementation will not pass my tests.

My understanding is that the `pull-window` implementation is not correct according to what I interpret debouncing to mean. Do you agree or am I misinterpreting what is really going on?

(re: https://css-tricks.com/the-difference-between-throttling-and-debouncing/#article-header-id-1)

**Comment by amsross** [comment]
@regular

**Comment by regular** [comment]
Hi, @amsross, sorry, I missed your previous post, thanks for pinging and thanks for taking the discussion here. Our understanding of what debounce is/should do is the same. I actually think that both our implementations have issues. I think yours might call the callback twice under certain circumstances, that's one reason why it does not pass the tests in this repo. I vaguely remember that my first implementation was somewhat like yours before I started using pull-window. I'll be diving deeper into this over the weekend. From what you wrote I understand you say my tests are not correct? Can you elaborate on that?

**Comment by amsross** [comment]
I'm poking both approaches right now and trying to clarify to myself what's going on. My approach only seems to work when the source is a `pull-pushable`, otherwise it ends up misbehaving in a couple of ways (including multiple callbacks). I'm trying to iron out exactly why that is.

Regarding what I said previously about the tests, here is an example of some values and the number of ms after start they were emitted:
```
value  ms
5      14
6      28
7      41
8      54
9      67  <-- 51ms after the value 5
10     84
           <-- 51ms gap
15     135
16     147
17     157
18     170
19     187 <-- 52ms after the value 15
20     198
```

My expectation would be that the only values that come through `debounce(50)` would be `[10, 20]`. Rather than waiting for silence for the given duration (`50ms`), the `pull-window` approach segments these values into 50ms windows so that the resulting values are `[9, 10, 19, 20]`. I don't think this is exposed in the existing tests.

Something like this:
```
  pull(
    pull.values([
      [10,  5],
      [10,  6],
      [10,  7],
      [10,  8],
      [10,  9],
      [10, 10],
      [50, 15],
      [10, 16],
      [10, 17],
      [10, 18],
      [10, 19],
      [10, 20],
    ]),
    pull.asyncMap(function(item, cb) {
      setTimeout(function() {
        cb(null, item[1])
      }, item[0]);
    })
    debounce(50),
    pull.collect((err, arr) => {
      t.equal(err, null);
      t.deepEqual(arr, [10, 20]);
      t.end();
    })
  );
```

**Comment by regular** [comment]
Hi @amsross, I'm currently also poking around in both out implementations. Among other things, I augmented your code with debug output including timing information and use `once` to prevent multiple calls to cb.

```
'use strict';

const once = require('once');

module.exports = function debounce (ms) {
  let timeout = null;
  let start = Date.now();

  return read => {
    return function next (end, cb) {
      cb = once(cb);
      read(end, function (end, data) {
        clearTimeout(timeout);
        let t = Date.now() - start;
        if (end) {
          console.log(t, data, 'calling cb because end');
          return cb(end, data);
        }

        timeout = setTimeout(() => {
          let t = Date.now() - start;
          console.log(t, data, 'calling cb (timeout)');
          cb(end, data);
        }, ms);
        return next(end, cb);
      });
    };
  };
};
```
You may want to try that out.

Regarding your comments about pull-window, I think that's a misconception. I agree that values should be emitted after a period of silence that exceeds the the specified time. That's exactly how the tests work. Look at the first one:

```
test.only('should ignore frequent updates', function(t) {
  pull(
    timedSource([
      [0,   0],
      [250, 1],
      [10,  2],
      [250, 3],
      [2000,4],
      [10,  5]
    ]),
    debounce(200),
    pull.through(console.log.bind(console)),
    pull.collect(function(end, arr) {
      t.notOk(end)
      t.deepEqual(arr, [0,2,3,5])
      console.log(arr)
      t.end();
    })
  })
})
```
Values 0, 2, 3 and 5 are emitted because they are followed by a gap of more than 200ms. My code passes that test, while yours falsely emits 0 and 1 and nothing else.

The test actually is equivalent to the test you proposed above, except that it does not require setTimeout to work as precise as is required for you rest to pass. setTimeout is not very precise, because the event loop might be blocked.

Unfortunately my code has a different problem though, it fail the "should pass through single item after timeout" test.

**Comment by amsross** [comment]
I think the key difference between those tests is that the window needs to be exceeded multiple times while consistently receiving values to expose what I was talking about.

I pushed up some changes while debugging last night. I was able to reliably pass a number of assertions, but I'm still seeing successes when the source is `pull-pushable` and failures when the source is buffered through `pull-asyncMap`. The fact that these two use cases are different is pretty perplexing to me. Maybe this relates to my sample data not actually being as equivalent as I think in `pushableSource` vs `asyncSource`?
https://github.com/amsross/pull-debounce/tree/fd9cebd1c81468012899f4ee2fd47fa4d6f22c54

I've been logging similar to your suggestion above, but not committing it:
```
diff --git index.js index.js
index 1e16476..ec0f5d5 100644
--- index.js
+++ index.js
@@ -8,12 +8,15 @@ module.exports = function debounce (ms) {
   return read => {
     return function next (end, cb) {
       read(end, function (end, data) {
+        console.log(`\t\t\t\tentered`, end, called, lastData, data);
         clearTimeout(timeout);
 
         if (end) {
           if (!called) {
+            console.log('setting late', lastData);
             timeout = setTimeout(() => {
               called = true;
+              console.log('sending late', lastData);
               cb(null, lastData);
             }, ms);
             return;
@@ -29,9 +32,21 @@ module.exports = function debounce (ms) {
           cb(null, data);
         }, ms);
 
+        console.log(`\t\t\t\tnexted`, end, called, lastData, data);
         next(end, cb);
         return;
       });
     };
   };
 };
+
+// const pull = require('pull-stream');
+// const window = require('pull-window');
+// module.exports = function(time) {
+//   return pull(
+//     window.recent(null, time || 1),
+//     pull.map(function (item) {
+//       return item.pop();
+//     });
+//   );
+// };
diff --git test/index.js test/index.js
index 815842f..eb7db6a 100644
--- test/index.js
+++ test/index.js
@@ -6,6 +6,8 @@ const debounce = require('../');
 const Pushable = require('pull-pushable');
 
 function pushableSource (x) {
+  const now = Date.now();
+
   const p = Pushable();
   let i = 1;
   const pusher = setInterval(() => {
@@ -21,11 +23,14 @@ function pushableSource (x) {
   return pull(
     p,
     pull.map(x => {
+      console.log(`\t\tpush`, x, Date.now() - now);
       return x;
     }));
 }
 
 function asyncSource (x) {
+  const now = Date.now();
+
   return pull(
     pull.values([
       [x, 1],
@@ -47,6 +52,7 @@ function asyncSource (x) {
     ]),
     pull.asyncMap(function (item, cb) {
       setTimeout(function () {
+        console.log(`\t\tasync`, item[1], Date.now() - now);
         cb(null, item[1]);
       }, item[0]);
     })
@@ -122,6 +128,7 @@ test('debounce returns all values when ms is less than data interval from async'
     asyncSource(10),
     debounce(5),
     pull.collect((err, arr) => {
+      console.log('END');
       t.equal(err, null, 'no error');
       t.deepEqual(arr, [
         1, 2, 3, 4, 5,

```

**Comment by regular** [comment]
Thinking about it, I think the main problem with your implementation is this: You have a race between setTimeout and read, if read wins the race, the timeout is cancelled, that's fine. However, if the timeout wins, the read is still pending and eventually will call the same callback again. That's wrong. Instead it should queue the value it read, return and wait for the client to call next with a new callback.

**Comment by regular** [comment]
> I think the key difference between those tests is that the window needs to be exceeded multiple times while consistently receiving values to expose what I was talking about.

I added your test to my test.js and wow, yes, I finally see what you mean now!

I am currently playing around with the idea of queueing when the timeout won.

**Comment by regular** [comment]
I have a solution that passes all the tests! Turns out it is even more complicated and requires some more state. Let me clean it up, then I push it.

**Comment by regular** [comment]
It's pushed and published as 1.1.1. If you have any ideas for simplification, that would be welcome. That was a nice collaboration, I enjoyed it, thanks!

**Comment by amsross** [comment]
Awesome work, and many thanks for the opportunity to help out!
